### PR TITLE
LoadGudrunOutput improvements

### LINKED ISSUE #23388: LoadGudrunOutput improvements
**Original reporter:** Helen Playford - ISIS

After the first implementation of `LoadGudrunOutput` some additional features should be added to build on the algorithm. The request features are:
* Load some of the meta data at a minimum this should using the run title as the workspace name
* Get the loader to examine expected file types as a group not individually - e.g. file dialogue will show all file extensions at once not individually

### PR DESCRIPTION
**Pull request by ElliotAOram**
**Description of work.**
* Added some UX changes:
  * Allow for multiple files to be viewed in the load dialog at a time
  * Get workspace name from loaded file meta data 

**Report to:** [Helen Playford - ISIS]

**To test:**
* Open the `LoadGudrunOutput` dialog box and click the browse button
* Locate the testing files in the build/ExternalData/Testing/Data/UnitTest directory
POLARIS00097947-min.{dcs01/mdcs01/mint01/mdor01/mgor01}
* Ensure that:
  * you can see all the valid file extensions in the browse menu.
  * you can filter what you see by changing the file extension from the drop down menu
* Select any of the workspaces and click `Run`
* Ensure that the output workspace has the same name as the top line of the input file (except `.` is replaced by `-` and no `# `)
 * Ensure that the workspace takes a custom name if you provide one. 

Fixes #23388. 

This does not require release notes because these are further changes to a new algorithm hence already covered sufficiently in the release notes

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by ElliotAOram** [comment]
**Note to tester:** Is there a cleaner way to deal with not providing an OutputWorkspace?
At the moment I have just provided a default value of `None` and then this will be overwritten if the user does not supply an OuputWorkspace name of anything else. As far as I could see from looking around at other places in the code base there didn't seem to be a much nicer way to do this. Perhaps just asking for a string as the OutputWorksace name is better, but then you lose the `Outputworkspace` as a `WorkspaceProperty` functionality.

**Comment by ElliotAOram** [comment]
Currently blocked by #23519

File: qt/widgets/common/src/FileDialogHandler.cpp [hunk 1]
```diff
@@ -161,6 +161,9 @@ QString getFilter(const std::vector<std::string> &exts) {
  */
 QString formatExtension(const std::string &extension) {
   QString formattedExtension = QString::fromStdString(extension);
+  if (extension == "") {
```
Comment (gvardany) [inline qt/widgets/common/src/FileDialogHandler.cpp]: `.empty()`

File: Framework/DataHandling/src/SaveOpenGenieAscii.cpp [hunk 2]
```diff
@@ -37,7 +37,7 @@ void SaveOpenGenieAscii::init() {
 
   // Declare required parameters, filename with ext {.his} and input
   // workspace
-  const std::vector<std::string> exts{".his", ".txt", ""};
+  const std::vector<std::string> exts{".his", ".txt"};
```
Comment (gvardany) [inline Framework/DataHandling/src/SaveOpenGenieAscii.cpp]: Can you explain these changes? It it intentional to forbid the files without extension, that were allowed before?

**Review by gvardany — CHANGES_REQUESTED** [review]
Few small remarks inline:

File: Framework/PythonInterface/plugins/algorithms/LoadGudrunOutput.py [hunk 3]
```diff
@@ -19,16 +19,37 @@ def summary(self):
...
         self.declareProperty(FileProperty(name='InputFile', defaultValue='', action=FileAction.Load,
-                                          extensions=['dcs01', 'mdcs01', 'mint01', 'mdor01', 'mgor01']))
-        self.declareProperty(WorkspaceProperty(name='OutputWorkspace', defaultValue='', direction=Direction.Output))
+                                          extensions=[".dcs01",
+                                                      ".mdsc01",
+                                                      ".mint01",
+                                                      ".mdor01",
+                                                      ".mgor01"]),
+                             doc="Gudrun output file to be loaded.")
+        self.declareProperty(WorkspaceProperty(name='OutputWorkspace', defaultValue='None',
```
Comment (gvardany) [inline Framework/PythonInterface/plugins/algorithms/LoadGudrunOutput.py]: Have you tried marking this as optional, with default value empty? I'm a bit hesitant about automatic generation of names; while it makes sense when running from dialog, it gets confusing when running from python scripts. If the output is mandatory it makes more sense to me to force typing a name for it.

**Review by gvardany — CHANGES_REQUESTED** [review]

File: Framework/DataHandling/src/SaveOpenGenieAscii.cpp (hunk 2 above)
Comment (ElliotAOram) [reply Framework/DataHandling/src/SaveOpenGenieAscii.cpp]: This was part of the working solution that I have accidentally committed! Thanks for picking this up I'll remove it now

File: Framework/PythonInterface/plugins/algorithms/LoadGudrunOutput.py (hunk 3 above)
Comment (ElliotAOram) [reply Framework/PythonInterface/plugins/algorithms/LoadGudrunOutput.py]: Much cleaner solution - thanks

**Comment by ElliotAOram** [comment]
@gvardany should be ready for review again now. Thanks for the first round of comments 👍

File: Framework/PythonInterface/plugins/algorithms/LoadGudrunOutput.py [hunk 4]
```diff
@@ -19,16 +20,38 @@ def summary(self):
...
+                                          extensions=[".dcs01",
+                                                      ".mdsc01",
+                                                      ".mint01",
+                                                      ".mdor01",
+                                                      ".mgor01"]),
+                             doc="Gudrun output file to be loaded.")
+        self.declareProperty(WorkspaceProperty(name='OutputWorkspace', defaultValue='',
+                                               direction=Direction.Output,
+                                               optional=PropertyMode.Optional),
+                             doc="If OutputWorkpsace is None,"
```
Comment (gvardany) [inline Framework/PythonInterface/plugins/algorithms/LoadGudrunOutput.py]: None -> empty?

**Review by gvardany — COMMENTED** [review]
Thanks for the changes, much better now! Just one more leftover:

**Review by gvardany — APPROVED** [review]

**Pull request merged**